ml-agents-envs/mlagents/envs/environment.py

```diff
@@ -279,156 +251,145 @@ def executable_launcher(self, file_name, docker_training, no_graphics, args):
+                    n_agents = self._env_state[group_name].n_agents()
+                self._env_actions[group_name] = self._empty_action(
+                    self._env_specs[group_name], n_agents
                 )
+        step_input = self._generate_step_input(self._env_actions)
+        with hierarchical_timer("communicator.exchange"):
+            outputs = self.communicator.exchange(step_input)
+        if outputs is None:
+            raise UnityCommunicationException("Communicator has stopped.")
+        self._update_group_specs(outputs)
+        rl_output = outputs.rl_output
+        self._update_state(rl_output)
+        self._env_actions.clear()
+
+    def get_agent_groups(self) -> List[str]:
+        return list(self._env_specs.keys())
+
+    def _assert_group_exists(self, agent_group: str) -> None:
+        if agent_group not in self._env_specs:
+            raise UnityActionException(
+                "The group {0} does not correspond to an existing agent group "
+                "in the environment".format(agent_group)
+            )
 
-                expected_discrete_size = n_agent * len(
-                    self._brains[brain_name].vector_action_space_size
+    def set_actions(self, agent_group: str, action: np.array) -> None:
+        self._assert_group_exists(agent_group)
+        if agent_group not in self._env_state:
+            return
+        spec = self._env_specs[agent_group]
+        expected_type = np.float32 if spec.is_action_continuous() else np.int32
+        expected_shape = (self._env_state[agent_group].n_agents(), spec.action_size)
+        if action.shape != expected_shape:
+            raise UnityActionException(
+                "The group {0} needs an input of dimension {1} but received input of dimension {2}".format(
+                    agent_group, expected_shape, action.shape
                 )
+            )
+        if action.dtype != expected_type:
+            action = action.astype(expected_type)
+        self._env_actions[agent_group] = action
 
-                continuous_check = (
-                    self._brains[brain_name].vector_action_space_type == "continuous"
+    def set_action_for_agent(
+        self, agent_group: str, agent_id: int, action: np.array
+    ) -> None:
+        self._assert_group_exists(agent_group)
+        if agent_group not in self._env_state:
+            return
+        spec = self._env_specs[agent_group]
+        expected_shape = (spec.action_size,)
+        if action.shape != expected_shape:
+            raise UnityActionException(
+                "The Agent {0} in group {1} needs an input of dimension {2} but received input of dimension {3}".format(
+                    agent_id, agent_group, expected_shape, action.shape
                 )
+            )
+        expected_type = np.float32 if spec.is_action_continuous() else np.int32
+        if action.dtype != expected_type:
+            action = action.astype(expected_type)
```

Comment:
raise instead?

Reply:
I would like to allow the user to use the LL-API directly. I think we should try make it work without the user having to specify a dtype every time if possible

Reply:
Sounds good